File: software/examples/javascript/ExampleScribble.js
var Tinkerforge = require('tinkerforge');
var GM = require('gm');
var getPixels = require('get-pixels');

var HOST = 'localhost';
var PORT = 4223;
var UID = '7xwQ9g'; // Change to your UID
var WIDTH = 128;
var HEIGHT = 64;

var ipcon = new Tinkerforge.IPConnection(); // Create IP connection
var oled = new Tinkerforge.BrickletOLED128x64(UID, ipcon); // Create device object
var originX = WIDTH / 2;
var originY = HEIGHT / 2;
var length = HEIGHT / 2 - 2;
var angle = 0;

function drawImage(oled, image) {
    // FIXME: GraphicsMagick doesn't seem to have a way to access the individual
    // pixels. Convert to PNG and then used get-pixels to read the pixels back.
    // This is far from ideal, but better than nothing.
    image.toBuffer('PNG', function (err, buffer) {
        if (err) {
            console.log('toBuffer: ' + err);
            return;
        }

        getPixels(buffer, 'image/png', function(err, pixels) {
            if (err) {
                console.log('getPixels: ' + err);
                return;
            }

            column = [];
            column_write = [];

            for (var i = 0; i < HEIGHT/8; i++) {
                column[i] = [];
            }

            for (var i = 0; i < HEIGHT/8; i++) {
                for (var j = 0; j < WIDTH; j++) {
                    page = 0;

                    for (var k = 0; k < 8; k++) {
                        if (pixels.get(j, i*8 + k, 0) > 0) {
                            page |= 1 << k;
                        }
                    }

                    column[i][j] = page;
                }
            }

            oled.newWindow(0, WIDTH-1, 0, 7)

            for (var i = 0; i < HEIGHT/8; i++) {
                l = 0;
                for (j = 0; j < WIDTH/2; j++) {
                    column_write[l] = column[i][j];
                    l++;
                }

                oled.write(column_write);

                l = 0;
                for (k = WIDTH/2; k < WIDTH; k++) {
                    column_write[l] = column[i][k];
                    l++;
                }

                oled.write(column_write);
            }
        });
    });
}

ipcon.connect(HOST, PORT,
    function (error) {
        console.log('Error: ' + error);
    }
); // Connect to brickd
// Don't use device before ipcon is connected

ipcon.on(Tinkerforge.IPConnection.CALLBACK_CONNECTED,
    function (connectReason) {
        // Clear display
        oled.clearDisplay();

        // Draw rotating line
        function render() {
            // FIXME: Creating the image once and reusing it would be better, but
            // on the second call toBuffer() fails with "Stream yields empty buffer"
            var image = GM(WIDTH, HEIGHT, '#000');
            var radians = angle * Math.PI / 180;
            var x = Math.floor(originX + length * Math.cos(radians));
            var y = Math.floor(originY + length * Math.sin(radians));

            image.fill('#FFF').drawLine(originX, originY, x, y);

            drawImage(oled, image);

            angle++;
        }

        setInterval(render, 25);
    }
);

console.log('Press key to exit');
process.stdin.on('data',
    function (data) {
        ipcon.disconnect();
        process.exit(0);
    }
);

File: software/examples/vbnet/ExampleScribble.vb
Imports System
Imports System.Drawing
Imports System.Math
Imports System.Threading
Imports Tinkerforge

Module ExampleScribble
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223
    Const UID As String = "XYZ" ' Change to your UID
    Const WIDTH As Integer = 128
    Const HEIGHT As Integer = 64

    Sub DrawImage(ByRef oled As BrickletOLED128x64, ByVal bitmap As Bitmap)
        Dim pages()() As Byte = New Byte(HEIGHT \ 8)() {}

        For row As Integer = 0 To HEIGHT \ 8 - 1
            pages(row) = New Byte(WIDTH) {}

            For column As Integer = 0 To WIDTH - 1
                pages(row)(column) = 0

                For bit As Integer = 0 To 7
                    If bitmap.GetPixel(column, (row * 8) + bit).GetBrightness() > 0 Then
                        pages(row)(column) = pages(row)(column) Or Convert.ToByte(1 << bit)
                    End If
                Next bit
            Next column
        Next row

        oled.NewWindow(0, WIDTH - 1, 0, HEIGHT \ 8 - 1)

        Dim section() As Byte = New Byte(64) {}

        For row As Integer = 0 To HEIGHT \ 8 - 1
            For column As Integer = 0 To WIDTH - 1 Step 64
                Array.Copy(pages(row), column, section, 0, 64)
                oled.Write(section)
            Next row
        Next row
    End Sub

    Sub Main()
        Dim ipcon As New IPConnection() ' Create IP connection
        Dim oled As New BrickletOLED128x64(UID, ipcon) ' Create device object

        ipcon.Connect(HOST, PORT) ' Connect to brickd
        ' Don't use device before ipcon is connected

        ' Clear display
        oled.ClearDisplay()

        ' Draw rotating line
        Dim bitmap As New Bitmap(WIDTH, HEIGHT)
        Dim originX As Integer = WIDTH \ 2
        Dim originY As Integer = HEIGHT \ 2
        Dim length As Integer = HEIGHT \ 2 - 2
        Dim angle As Integer = 0

        Console.WriteLine("Press enter to exit")

        While Not Console.KeyAvailable
            Dim radians As Double = Math.PI * angle / 180.0
            Dim x As Integer = Convert.ToInt32(Math.Floor(originX + length * Math.Cos(radians)))
            Dim y As Integer = Convert.ToInt32(Math.Floor(originY + length * Math.Sin(radians)))

            Using g As Graphics = Graphics.FromImage(bitmap)
                g.FillRectangle(Brushes.Black, 0, 0, WIDTH, HEIGHT)
                g.DrawLine(Pens.White, originX, originY, x, y)
            End Using

            DrawImage(oled, bitmap)
            Thread.Sleep(25)

            angle += 1
        End While

        Console.WriteLine("Press key to exit")
        Console.ReadLine()
        ipcon.Disconnect()
    End Sub
End Module
